### PR TITLE
dev/msp: cron job, private images support

### LINKED ISSUE #57100: msp: cron job support
e.g. [Cody Gateway usage exporter](https://sourcegraph.slack.com/archives/C05497E9MDW/p1695838494483179)

 This should include the availability of similar runtime utilities/toolkits to long-running MSP services.

### PR DESCRIPTION
**Pull request by bobheadxi**
This change:

1. Adds support for `kind: job` services, and a new specification field `cron:` for cron-job-related options - see demo in https://github.com/sourcegraph/managed-services/pull/42. Closes #57100 
2. Adds automatic support for private image registries operated by Sourcegraph in GCP - when you provide a `build.image` that points to a GCR or Artifact Registry image, we automatically grant the Cloud Run robot account various object/artifact-reading permissions to the target GCP project specified in the image repo name.

A few larger refactors were needed to make this happen:

1. In `dev/managedservicesplatform/internal/stack/cloudrun`, we now have 2 internal packages that implement an abstraction for a "Cloud Run resource builder", so that we can generalize the configuration of both services and jobs
2. Service-specific and job-specific configuration now lives in embedded structs that group them by purpose, and more configuration options are now nil-able so that we can have configuration exclusive to services/jobs

## Test plan

- [x] https://github.com/sourcegraph/managed-services/pull/45 applies changes to existing services - see minimal diff on manual review
- [x] https://github.com/sourcegraph/managed-services/pull/42 is a test run of a job service: 
Screenshot 2023-11-06 at 5 44 59 PM
Screenshot 2023-11-06 at 5 45 14 PM

### CODE REVIEW

**Comment by bobheadxi** [comment]
Current dependencies on/for this PR:
* main
  * **PR #57930** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/57930" target="_blank">Graphite</a>  👈
    * **PR #58131** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/58131" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/57930?utm_source=stack-comment).

File: dev/managedservicesplatform/internal/stack/cloudrun/cloudrun.go [hunk 1]
```diff
@@ -132,8 +122,73 @@ func NewStack(stacks *stack.Set, vars Variables) (*Output, error) {
...
+			cloudRunDependencies = append(cloudRunDependencies,
+				projectiammember.NewProjectIamMember(stack,
+					id.ResourceID("member_image_access_%s", r.ID),
+					&projectiammember.ProjectIamMemberConfig{
+						Project: imageProject,
+						Role:    pointers.Ptr(r.Role),
+						Member: pointers.Ptr(fmt.Sprintf("serviceAccount:%s",
+							vars.CloudRunIdentityEmail)),
+					}))
+		}
```
Comment (michaellzc) [inline dev/managedservicesplatform/internal/stack/cloudrun/cloudrun.go:151]: Project level iam permission feels too broad to me.

For artifact registry, it's possible to grant repository-level permission. not sure about GCR. If it doesn't, maybe we can treat them differently? It's quite easy to distinguish between gcr and artifact registry

**Review by michaellzc — COMMENTED** [review]
overal lgtm

File: dev/managedservicesplatform/internal/stack/cloudrun/cloudrun.go (hunk 1 above)
Comment (bobheadxi) [reply dev/managedservicesplatform/internal/stack/cloudrun/cloudrun.go:151]: Good point, but I think I might leave this as-is for now, though if you object strongly I'm happy to follow up after this PR :) The robot account we grant access to is only used for deploying services to Cloud Run, so I think this is a fairly low-risk vector, and for the most part, the status quo has been to grant project-wide image access (example: https://github.com/sourcegraph/infrastructure/pull/5382)

Comment (michaellzc) [reply dev/managedservicesplatform/internal/stack/cloudrun/cloudrun.go:151]: that’s fine, not blocking!

**Review by michaellzc — APPROVED** [review]
lgtm

**Pull request merged**